### PR TITLE
txn: generation in pipelined DML

### PR DESCRIPTION
**Pull request by ekexium**
### What is changed and how it works?

Issue Number: ref #16291

What's Changed:

```commit-message
Introduce the generation in Pipelined-DML.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
 None
```

### CODE REVIEW

File: src/storage/txn/actions/prewrite.rs [hunk 1]
```diff
@@ -41,6 +41,28 @@ pub fn prewrite<S: Snapshot>(
     secondary_keys: &Option<Vec<Vec<u8>>>,
     pessimistic_action: PrewriteRequestPessimisticAction,
     expected_for_update_ts: Option<TimeStamp>,
+) -> Result<(TimeStamp, OldValue)> {
+    prewrite_with_generation(
```
Comment (cfzjywxk) [inline src/storage/txn/actions/prewrite.rs:45]: How about adding `generation` as an input parameter of `prewrite` function, if it's `None` related procesing could be skipped.

Comment (ekexium) [reply src/storage/txn/actions/prewrite.rs:45]: That's my first thought too. But it exposes `generation` to many places where the users of this function don't have to know about it. By this wrapper we hide that information for most of the usages.
Besides, it saves a lot of modifications in those places, especially in tests.

File: components/file_system/src/io_stats/mod.rs [hunk 2]
```diff
@@ -53,6 +53,7 @@ pub use proc::*;
 // See https://man7.org/linux/man-pages/man2/open.2.html#NOTES
 #[cfg(test)]
 #[repr(align(512))]
+#[cfg_attr(not(target_os = "linux"), allow(unused))]
```
Comment (cfzjywxk) [inline components/file_system/src/io_stats/mod.rs:56]: Unrelated change?

File: components/txn_types/src/lock.rs [hunk 3]
```diff
@@ -713,7 +737,7 @@ mod tests {
     }
 
     #[test]
-    fn test_lock() {
+    fn mtest_lock() {
```
Comment (cfzjywxk) [inline components/txn_types/src/lock.rs:740]: Unrelated change?

File: components/txn_types/src/lock.rs [hunk 4]
```diff
@@ -910,7 +934,8 @@ mod tests {
                 false,
             )
             .set_last_change(LastChange::make_exist(4.into(), 2))
-            .set_txn_source(1),
+            .set_txn_source(1)
+            .with_generation(10),
```
Comment (cfzjywxk) [inline components/txn_types/src/lock.rs:938]: The result seems not to be checked.

File: deny.toml [hunk 5]
```diff
@@ -106,4 +106,4 @@ exceptions = [
 [sources]
 unknown-git = "deny"
 unknown-registry = "deny"
-allow-org = { github = ["tikv", "pingcap", "rust-lang"] }
+allow-org = { github = ["tikv", "pingcap", "rust-lang", "ekexium"] }
```
Comment (cfzjywxk) [inline deny.toml:109]: Why change this?

File: src/storage/txn/actions/prewrite.rs [hunk 6]
```diff
@@ -408,6 +437,14 @@ impl<'a> PrewriteMutation<'a> {
             return Ok(LockStatus::Pessimistic(lock.for_update_ts));
         }
 
+        if generation_to_write > 0 && lock.generation >= generation_to_write {
```
Comment (you06) [inline src/storage/txn/actions/prewrite.rs:440]: Can we skip this key when `lock.generation == generation_to_write`, imaging this case:

1. flush success but the response1 is slow.
2. retried flush got generation-out-of-order error.

In this case, client know the retried flush request is failed, but nothing can do except waiting for response of the first flush request, if this response is lost, the transaction has to be fail.

File: components/file_system/src/io_stats/mod.rs (hunk 2 above)
Comment (ekexium) [reply components/file_system/src/io_stats/mod.rs:56]: This is to fix the clippy warning on non-linux targets.

File: components/txn_types/src/lock.rs (hunk 3 above)
Comment (ekexium) [reply components/txn_types/src/lock.rs:740]: oops

File: components/txn_types/src/lock.rs [hunk 7]
```diff
@@ -1203,7 +1263,7 @@ mod tests {
             for_update_ts: TimeStamp(101), txn_size: 10, min_commit_ts: TimeStamp(127), \
             use_async_commit: true, secondaries: [], rollback_ts: [], \
             last_change: Exist { last_change_ts: TimeStamp(80), estimated_versions_to_last_change: 4 }, txn_source: 0\
-            , is_locked_with_conflict: false }"
+            , is_locked_with_conflict: false, generation: 0 }"
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/txn_types/src/lock.rs:1266]: makes sense

File: components/error_code/src/storage.rs [hunk 8]
```diff
@@ -45,6 +45,7 @@ define_error_codes!(
 
     PRIMARY_MISMATCH => ("PrimaryMismatch", "", ""),
     UNDETERMINED => ("Undetermined", "", ""),
+    GENERATION_OUT_OF_ORDER => ("GenerationOutOfOrder", "", ""),
```
[the remark this answers is not in the document]
Comment (ekexium) [reply components/error_code/src/storage.rs:48]: It makes sense not to return the error to client.

File: src/storage/txn/actions/prewrite.rs (hunk 6 above)
Comment (ekexium) [reply src/storage/txn/actions/prewrite.rs:440]: Let's simply skip all mutations with `GenerationOutOfOrder` errors.

File: src/storage/txn/commands/flush.rs [hunk 9]
```diff
@@ -180,6 +188,16 @@ impl Flush {
...
                     }
                 }
+                Err(crate::storage::mvcc::Error(
+                    box crate::storage::mvcc::ErrorInner::GenerationOutOfOrder(generation, lock),
+                )) => {
+                    info!(
+                        "generation in Flush is smaller than that in lock, ignore this mutation";
+                        "start_ts" => self.start_ts,
+                        "generation" => generation,
+                        "lock" => ?lock,
```
Comment (cfzjywxk) [inline src/storage/txn/commands/flush.rs:198]: Should the log be printted using red-act interfaces?

**Review by cfzjywxk — APPROVED** [review]
LGTM

File: src/storage/txn/commands/flush.rs (hunk 9 above)
Comment (ekexium) [reply src/storage/txn/commands/flush.rs:203]: Right. I've changed it from `LockInfo` to `Lock` and `Key`, which respects redact setting

**Comment by cfzjywxk** [comment]
@you06 
PTAL

File: components/error_code/src/storage.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (MyonKeminta) [reply components/error_code/src/storage.rs:48]: There's a benefit returning the error to the client: if the client sends out-of-order value due to bugs, it will receive the error instead of continuing its incorrect behavior silently.

[the remark this answers is not in the document]
Comment (ekexium) [reply components/error_code/src/storage.rs:48]: > There's a benefit returning the error to the client: if the client sends out-of-order value due to bugs, it will receive the error instead of continuing its incorrect behavior silently.

But the only thing we can do in client side is still ignoring the error?

[the remark this answers is not in the document]
Comment (you06) [reply components/error_code/src/storage.rs:48]: This error should be dropped in the kv client in the common path:

1. requests with generation n sent
2. requests with generation n retried sending
3. requests with generation n done
4. requests with generation n+1 sent
5. `GenerationOutOfOrder` order responsed, ignored by client, because the requests are already canceled.

If the geneartion in client jumped back due to bugs, it may receive this error.

1. requests with generation n sent
2. requests with generation n retried sending
3. requests with generation n done
4. requests with generation n-1 sent
5. client got `GenerationOutOfOrder`

**Review by you06 — APPROVED** [review]
